src/relay/pass/type_infer.cc

```diff
@@ -22,607 +22,360 @@
+//
+class TypeInferencer : private ExprFunctor<Type(const Expr&)> {
+ public:
+  // constructors
+  TypeInferencer() {
   }
-
-  TypeRelation ToTypeRel() const {
-    Array<Type> args = Array<Type>(this->args.begin(), this->args.end());
-    return TypeRelationNode::make(
-        this->name, this->func, args);
+  explicit TypeInferencer(Environment env)
+      : env_(env) {
   }
-};
-
-struct TypeContext {
-  std::unordered_map<Var, Type, NodeHash> var_map;
-  std::vector<std::vector<TypeRelationData> > constraints;
 
-  TypeContext() { constraints.push_back({}); }
+  // inference the type of expr.
+  Expr Infer(Expr expr);
 
-  void Insert(const Var& id, const Type& t) { var_map[id] = t; }
-
-  void AddConstraint(const TypeConstraint& constraint) {
-      constraints.back().push_back(TypeRelationData(Downcast<TypeRelation>(constraint)));
+ private:
+  // type resolver that maps back to type
+  class Resolver;
+  // internal environment
+  Environment env_;
+  // map from expression to checked type
+  // type inferencer will populate it up
+  std::unordered_map<Expr, Type, NodeHash, NodeEqual> type_map_;
+  // The solver used by the inferencer.
+  TypeSolver solver_;
+  // Unify two types
+  Type Unify(const Type& t1, const Type& t2, const Span& span) {
+    // TODO(tqchen, jroesch): propagate span to solver
+    try {
+      return solver_.Unify(t1, t2);
+    } catch (const dmlc::Error &e) {
+      LOG(FATAL)
+          << "Error unifying `"
+          << t1
+          << "` and `"
+          << t2
+          << "`: " << e.what();
+      return Type();
+    }
   }
-
-  Type Lookup(const Var& var) {
-    auto type = var_map.find(var);
-    if (type != var_map.end()) {
-      return (*type).second;
-    } else {
-      throw FatalTypeError(std::string("undeclared local variable: ") + var->name_hint);
+  // Lazily get type for expr
+  // will call visit to deduce it if it is not in the type_map_
+  Type GetType(const Expr &expr) {
+    auto it = type_map_.find(expr);
+    if (it != type_map_.end()) {
+      return it->second;
     }
+    Type ret = this->VisitExpr(expr);
+    type_map_[expr] = ret;
+    return ret;
   }
 
-  struct Scope {
-    TypeContext& tc;
-    explicit Scope(TypeContext& tc) : tc(tc) { tc.constraints.push_back({}); }
-    ~Scope() { tc.constraints.pop_back(); }
-  };
-};
-
-struct CheckedExpr {
-  Expr expr;
-  Type type;
-  CheckedExpr(Expr e, Type t) : expr(e), type(t) {}
-  CheckedExpr() {}
-};
-
-enum SolverResult : int;
-
-class TypeInferencer : private ExprFunctor<CheckedExpr(const Expr&)> {
- private:
-  TypeContext context;
-
- public:
-  Environment env;
-  TypeUnifier unifier;
-
-  template <typename T>
-  T WithScope(const std::function<T()>& f) {
-    TypeContext::Scope fr(context);
-    return f();
+  // Visitor logics
+  Type VisitExpr_(const VarNode* op) final {
+    // The type of Var can already been lookedup in type_map_;
+    LOG(FATAL) << "Cannot find binding for var " << GetRef<Var>(op);
```

Comment:
I feel like we should still keep an error reporting interface which is span focused. The problem here is that the span on the variable is meaningless due to it's underlying container being contained in many places in the expression.